compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ClassCodegen.kt

```diff
@@ -161,33 +164,46 @@ open class ClassCodegen protected constructor(
             visitor.visitSource(shortName, null)
         }
 
-        // Delay generation of <clinit> until the end because inline function calls
-        // might need to generate the `$assertionsDisabled` field initializer.
-        classInitializer = irClass.functions.singleOrNull { it.name.asString() == "<clinit>" }
-        for (declaration in irClass.declarations) {
-            if (declaration != classInitializer)
-                generateDeclaration(declaration)
+        // JVMS7 (4.7.6): a nested class or interface member will have InnerClasses information
+        // for each enclosing class and for each immediate member
+        parentClassCodegen?.let { writeInnerClass(irClass, typeMapper, context, it.visitor) }
+        for (codegen in generateSequence(this) { it.parentClassCodegen }.takeWhile { it.parentClassCodegen != null }) {
+            writeInnerClass(codegen.irClass, typeMapper, context, visitor)
         }
-        classInitializer?.let {
-            generatingClInit = true
-            generateMethod(it)
+
+        // JVMS7 (4.7.7): A class must have an EnclosingMethod attribute if and only if
+        // it is a local class or an anonymous class.
+        //
+        // The attribute contains the innermost class that encloses the declaration of
+        // the current class. If the current class is immediately enclosed by a method
+        // or constructor, the name and type of the function is recorded as well.
+        if (parentClassCodegen != null) {
+            val enclosingFunction = context.customEnclosingFunction[irClass.attributeOwnerId] ?: parentFunction
+            if (enclosingFunction != null || irClass.isAnonymousObject) {
+                val method = enclosingFunction?.let(context.methodSignatureMapper::mapAsmMethod)
+                visitor.visitOuterClass(parentClassCodegen.type.internalName, method?.name, method?.descriptor)
+            }
         }
 
-        // Generate nested classes at the end, to ensure that codegen for companion object will have the necessary JVM signatures in its
-        // trace for properties moved to the outer class
-        for (codegen in nestedClasses) {
-            codegen.generate()
+        if (sourceMapShouldBeWritten) {
```

Comment:
Could you explaint a bit intention of moving this part from done?

Reply:
A separate done() is simply no longer necessary as coroutine-related code in FunctionCodegen calls generate() instead after the method transformer finishes.

Reply:
Actually `generate` method is too big, maybe extract inner class related logic to separate method?

Reply:
Sure.